src/ir/proto_to_ir.cc

```diff
@@ -0,0 +1,260 @@
+#include "src/ir/attributes/string_attribute.h"
+
+namespace raksha::ir {
+
+std::unique_ptr<Module> ProtoToIR::PreVisit(const proto::Module& module_proto) {
+  auto module = std::make_unique<Module>();
+  for (const auto& block_it : module_proto.blocks()) {
+    blocks_.insert({block_it.id(), BlockBuilder()});
+    PreVisit(block_it.block());
+  }
+  return module;
+}
+
+void ProtoToIR::PreVisit(const proto::Block& block_proto) {
+  for (const auto& operation_it : block_proto.operations()) {
+    const auto& operation_proto = operation_it.operation();
+    const auto& name = operation_proto.operator_name();
+    const Operator* op = context_.GetOperator(name);
+    CHECK(op != nullptr) << "Operator " << name << " was not registered.";
+
+    // Register operation result values with 'SsaNames'.
+    auto operation = std::make_unique<Operation>(*op);
+    value::OperationResult operation_result(*operation, "out");
+    auto operation_result_value = Value(operation_result);
+    ssa_names_.GetOrCreateID(operation_result_value);
+    operations_.insert({operation_it.id(), std::move(operation)});
+    PreVisit(operation_proto);
+  }
+  for (const auto& result_it : block_proto.results().values()) {
+    PreVisit(result_it.second);
+  }
+}
+
+void ProtoToIR::PreVisit(const proto::Operation& operation_proto) {
+  for (const auto& input_proto : operation_proto.inputs()) {
+    PreVisit(input_proto);
+  }
+  for (const auto& attribute_it : operation_proto.attributes().attributes()) {
+    PreVisit(attribute_it.second);
+  }
+  // TODO: PreVisit(impl_module);
+}
+
+void ProtoToIR::PreVisit(const proto::Value& value_proto) {
+  if (value_proto.has_any_value()) {
+    PreVisit(value_proto.any_value());
+  } else if (value_proto.has_operation_result_value()) {
+    PreVisit(value_proto.operation_result_value());
+  } else if (value_proto.has_block_argument_value()) {
+    PreVisit(value_proto.block_argument_value());
+  } else {
+    CHECK(false) << "Unexpected variant of Value from IR";
+  }
+}
+
+void ProtoToIR::PreVisit(const proto::AnyValue&) {
+  // Any is empty.
+}
+
+void ProtoToIR::PreVisit(
+    const proto::BlockArgumentValue& block_argument_proto) {
+  // Nothing to do here.
+}
+
+void ProtoToIR::PreVisit(
+    const proto::OperationResultValue& operation_argument_proto) {
+  // Nothing to do here.
+}
+
+void ProtoToIR::PreVisit(const proto::AttributePayload& attribute_payload) {
+  // Nothing to do here.
+}
+
+void ProtoToIR::PostVisit(Module& module, const proto::Module& module_proto) {
+  for (const auto& block_it : module_proto.blocks()) {
+    auto find_block_ptr = blocks_.find(block_it.id());
+    CHECK(find_block_ptr != blocks_.end())
+        << "This block builder should be pre-constructed.";
+    auto& builder = find_block_ptr->second;
+    PostVisit(builder, block_it.block());
+    module.AddBlock(builder.build());
+  }
+  // Register the module with 'SsaNames'.
+  ssa_names_.GetOrCreateID(module);
```

Comment:
Is there chance that this will be called multiple times for the same block? If not, you should be able to use `AddID`?

Reply:
Ah, good catch! Thanks